File: rust/src/git/commands.rs
use anyhow::{Context, Result, bail};
use std::path::{Path, PathBuf};
use std::process::Command;

/// Result of running a git command.
#[derive(Debug)]
pub struct GitResult {
    pub success: bool,
    pub stdout: String,
    pub stderr: String,
    pub exit_code: i32,
}

/// Run a git command and return the result.
pub fn git(args: &[&str], cwd: &str) -> Result<GitResult> {
    let output = Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .with_context(|| format!("failed to run git {}", args.join(" ")))?;

    Ok(GitResult {
        success: output.status.success(),
        stdout: String::from_utf8_lossy(&output.stdout).trim().to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).trim().to_string(),
        exit_code: output.status.code().unwrap_or(-1),
    })
}

/// Run a git command and return stdout, failing if the command fails.
pub fn git_stdout(args: &[&str], cwd: &str) -> Result<String> {
    let result = git(args, cwd)?;
    if !result.success {
        bail!(
            "git {} failed (exit {}): {}",
            args.join(" "),
            result.exit_code,
            result.stderr
        );
    }
    Ok(result.stdout)
}

/// Find the git root directory.
pub fn find_git_root(cwd: &str) -> Result<String> {
    git_stdout(&["rev-parse", "--show-toplevel"], cwd)
}

/// Workspace/monorepo manager config files, in precedence order matching workspace-tools.
pub const MANAGER_FILES: &[&str] = &[
    "lerna.json",
    "rush.json",
    "yarn.lock",
    "pnpm-workspace.yaml",
    "package-lock.json",
];

/// Walk up the directory tree from `cwd` looking for any of the given files.
/// Returns the full path of the first match, or None if not found.
pub fn search_up(files: &[&str], cwd: &str) -> Option<PathBuf> {
    let mut dir = Path::new(cwd).to_path_buf();
    if let Ok(abs) = std::fs::canonicalize(&dir) {
        dir = abs;
    }

    loop {
        for f in files {
            let candidate = dir.join(f);
            if candidate.exists() {
                return Some(candidate);
            }
        }
        if !dir.pop() {
            break;
        }
    }
    None
}

/// Find the project root by searching up for workspace manager config files,
/// falling back to the git root. Matches workspace-tools findProjectRoot.
pub fn find_project_root(cwd: &str) -> Result<String> {
    if let Some(found) = search_up(MANAGER_FILES, cwd) {
        if let Some(parent) = found.parent() {
            return Ok(parent.to_string_lossy().to_string());
        }
    }
    find_git_root(cwd)
}

/// Get the current branch name.
pub fn get_branch_name(cwd: &str) -> Result<Option<String>> {
    let result = git(&["rev-parse", "--abbrev-ref", "HEAD"], cwd)?;
    if result.success {
        Ok(Some(result.stdout))
    } else {
        Ok(None)
    }
}

/// Get the user's email from git config.
pub fn get_user_email(cwd: &str) -> Option<String> {
    git_stdout(&["config", "user.email"], cwd).ok()
}

/// Get files changed between the current branch and the target branch.
pub fn get_branch_changes(branch: &str, cwd: &str) -> Result<Vec<String>> {
    let result = git(
        &[
            "--no-pager",
            "diff",
            "--name-only",
            "--relative",
            &format!("{branch}..."),
        ],
        cwd,
    )?;
    if !result.success {
        return Ok(vec![]);
    }
    Ok(process_git_output(&result.stdout))
}

/// Get staged changes.
pub fn get_staged_changes(cwd: &str) -> Result<Vec<String>> {
    let result = git(
        &[
            "--no-pager",
            "diff",
            "--staged",
            "--name-only",
            "--relative",
        ],
        cwd,
    )?;
    if !result.success {
        return Ok(vec![]);
    }
    Ok(process_git_output(&result.stdout))
}

/// Get changes between two refs, optionally filtering by pattern and diff filter.
pub fn get_changes_between_refs(
    from_ref: &str,
    diff_filter: Option<&str>,
    pattern: Option<&str>,
    cwd: &str,
) -> Result<Vec<String>> {
    let diff_flag = diff_filter.map(|f| format!("--diff-filter={f}"));
    let range = format!("{from_ref}...");
    let mut args: Vec<&str> = vec!["--no-pager", "diff", "--name-only", "--relative"];
    if let Some(ref flag) = diff_flag {
        args.push(flag);
    }
    args.push(&range);
    if let Some(pat) = pattern {
        args.push("--");
        args.push(pat);
    }

    let result = git(&args, cwd)?;
    if !result.success {
        return Ok(vec![]);
    }
    Ok(process_git_output(&result.stdout))
}

/// Get untracked files.
pub fn get_untracked_changes(cwd: &str) -> Result<Vec<String>> {
    let result = git(&["ls-files", "--others", "--exclude-standard"], cwd)?;
    Ok(process_git_output(&result.stdout))
}

/// Stage files.
pub fn stage(patterns: &[&str], cwd: &str) -> Result<()> {
    let mut args = vec!["add"];
    args.extend(patterns);
    git_stdout(&args, cwd)?;
    Ok(())
}

/// Commit with a message. Extra options can be passed (e.g. --only path).
pub fn commit(message: &str, cwd: &str, extra_options: &[&str]) -> Result<()> {
    let mut args = vec!["commit", "-m", message];
    args.extend(extra_options);
    git_stdout(&args, cwd)?;
    Ok(())
}

/// Check if a ref exists.
pub fn rev_parse_verify(reference: &str, cwd: &str) -> bool {
    git(&["rev-parse", "--verify", reference], cwd)
        .map(|r| r.success)
        .unwrap_or(false)
}

/// Check if the repository is a shallow clone.
pub fn is_shallow_repository(cwd: &str) -> bool {
    git_stdout(&["rev-parse", "--is-shallow-repository"], cwd)
        .map(|s| s == "true")
        .unwrap_or(false)
}

/// Find the merge base of two refs.
pub fn merge_base(ref1: &str, ref2: &str, cwd: &str) -> Result<Option<String>> {
    let result = git(&["merge-base", ref1, ref2], cwd)?;
    if result.success {
        Ok(Some(result.stdout))
    } else {
        Ok(None)
    }
}

/// Parse a remote branch string like "origin/main" into (remote, branch).
pub fn parse_remote_branch(branch: &str) -> Option<(String, String)> {
    let slash_pos = branch.find('/')?;
    Some((
        branch[..slash_pos].to_string(),
        branch[slash_pos + 1..].to_string(),
    ))
}

/// Returns the default remote name, matching workspace-tools getDefaultRemote.
///
/// The order of preference is:
/// 1. If `repository` is defined in package.json at the git root, the remote with a matching URL
/// 2. "upstream" if defined
/// 3. "origin" if defined
/// 4. The first defined remote
/// 5. "origin" as final fallback
///
/// Note: ADO/VSO URL formats are not currently handled by `get_repository_name`.
/// This is probably fine since usage of forks with ADO is uncommon.
pub fn get_default_remote(cwd: &str) -> String {
    let git_root = match find_git_root(cwd) {
        Ok(root) => root,
        Err(_) => return "origin".to_string(),
    };

    // Read package.json at git root for repository field
    let mut repository_name = String::new();
    let package_json_path = Path::new(&git_root).join("package.json");
    match std::fs::read_to_string(&package_json_path) {
        Err(_) => {
            crate::log_info!(
                r#"Valid "repository" key not found in "{}". Consider adding this info for more accurate git remote detection."#,
                package_json_path.display()
            );
        }
        Ok(contents) => {
            if let Ok(pkg) = serde_json::from_str::<serde_json::Value>(&contents) {
                let repo_url = match pkg.get("repository") {
                    Some(serde_json::Value::String(s)) => s.clone(),
                    Some(obj) => obj
                        .get("url")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string(),
                    None => String::new(),
                };
                if repo_url.is_empty() {
                    crate::log_info!(
                        r#"Valid "repository" key not found in "{}". Consider adding this info for more accurate git remote detection."#,
                        package_json_path.display()
                    );
                } else {
                    repository_name = get_repository_name(&repo_url);
                }
            } else {
                crate::log_info!(
                    r#"Valid "repository" key not found in "{}". Consider adding this info for more accurate git remote detection."#,
                    package_json_path.display()
                );
            }
        }
    }

    // Get remotes with URLs
    let remotes_result = match git(&["remote", "-v"], cwd) {
        Ok(r) if r.success => r,
        _ => return "origin".to_string(),
    };

    // Build mapping from repository name → remote name
    let mut remotes_by_repo_name: std::collections::HashMap<String, String> =
        std::collections::HashMap::new();
    let mut all_remote_names: Vec<String> = Vec::new();
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();

    for line in remotes_result.stdout.lines() {
        let fields: Vec<&str> = line.split_whitespace().collect();
        if fields.len() < 2 {
            continue;
        }
        let remote_name = fields[0];
        let remote_url = fields[1];
        let repo_name = get_repository_name(remote_url);
        if !repo_name.is_empty() {
            remotes_by_repo_name.insert(repo_name, remote_name.to_string());
        }
        if seen.insert(remote_name.to_string()) {
            all_remote_names.push(remote_name.to_string());
        }
    }

    // 1. Match by repository name from package.json
    if !repository_name.is_empty() {
        if let Some(matched) = remotes_by_repo_name.get(&repository_name) {
            return matched.clone();
        }
    }

    // 2-4. Fall back to upstream > origin > first
    if all_remote_names.iter().any(|r| r == "upstream") {
        return "upstream".to_string();
    }
    if all_remote_names.iter().any(|r| r == "origin") {
        return "origin".to_string();
    }
    if let Some(first) = all_remote_names.first() {
        return first.clone();
    }

    "origin".to_string()
}

/// Extracts the "owner/repo" full name from a git URL.
/// Handles HTTPS, SSH, git://, and shorthand (github:owner/repo) formats.
///
/// Note: Azure DevOps and Visual Studio Online URL formats are not currently handled.
/// Those would require more complex parsing similar to workspace-tools' git-url-parse usage.
pub fn get_repository_name(raw_url: &str) -> String {
    if raw_url.is_empty() {
        return String::new();
    }

    // SSH format: git@github.com:owner/repo.git or user@host:path
    if let Some(colon_pos) = raw_url.find(':') {
        if raw_url[..colon_pos].contains('@') && !raw_url[..colon_pos].contains('/') {
            let path = &raw_url[colon_pos + 1..];
            // Skip if path starts with / (would be ssh://user@host/path, not SCP syntax)
            if !path.starts_with('/') {
                return path.trim_end_matches(".git").to_string();
            }
        }
    }

    // Shorthand format: github:owner/repo (scheme without //)
    if let Some(colon_pos) = raw_url.find(':') {
        let scheme = &raw_url[..colon_pos];
        let rest = &raw_url[colon_pos + 1..];
        if scheme.chars().all(|c| c.is_ascii_lowercase())
            && !rest.starts_with("//")
            && !rest.is_empty()
        {
            return rest.trim_end_matches(".git").to_string();
        }
    }

    // HTTPS or git:// format — parse as URL
    if let Some(path_start) = raw_url.find("://") {
        let after_scheme = &raw_url[path_start + 3..];
        // Find the path part (after host)
        if let Some(slash_pos) = after_scheme.find('/') {
            let path = &after_scheme[slash_pos + 1..];
            let path = path.trim_end_matches(".git");
            if !path.is_empty() {
                return path.to_string();
            }
        }
    }

    String::new()
}

/// Get the default remote branch (tries to detect from git remote).
/// If `branch` is Some, combines it with the default remote.
/// If `branch` is None, detects the remote's HEAD branch, falling back to
/// git config init.defaultBranch or "master".
pub fn get_default_remote_branch(cwd: &str) -> Result<String> {
    let remote = get_default_remote(cwd);

    // Try to detect HEAD branch from remote
    if let Ok(r) = git(&["remote", "show", &remote], cwd) {
        if r.success {
            for line in r.stdout.lines() {
                let trimmed = line.trim();
                if let Some(rest) = trimmed.strip_prefix("HEAD branch:") {
                    let branch = rest.trim();
                    return Ok(format!("{remote}/{branch}"));
                }
            }
        }
    }

    // Fallback: try git config init.defaultBranch
    if let Ok(default_branch) = git_stdout(&["config", "init.defaultBranch"], cwd) {
        return Ok(format!("{remote}/{default_branch}"));
    }

    Ok(format!("{remote}/master"))
}

/// Get the default remote branch with a specific branch name.
/// Combines the branch with the default remote name.
pub fn get_default_remote_branch_for(cwd: &str, branch: &str) -> String {
    let remote = get_default_remote(cwd);
    format!("{remote}/{branch}")
}

/// List all tracked files matching a pattern.
pub fn list_tracked_files(pattern: &str, cwd: &str) -> Result<Vec<String>> {
    let result = git(&["ls-files", pattern], cwd)?;
    Ok(process_git_output(&result.stdout))
}

/// Fetch a branch from a remote.
pub fn fetch(remote: &str, branch: &str, cwd: &str, depth: Option<u32>) -> Result<()> {
    let mut args = vec!["fetch".to_string(), "--".to_string(), remote.to_string()];
    if let Some(d) = depth {
        args.push(format!("--depth={d}"));
    }
    args.push(branch.to_string());

    let str_args: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
    let result = git(&str_args, cwd)?;
    if !result.success {
        bail!(
            "Fetching branch \"{branch}\" from remote \"{remote}\" failed: {}",
            result.stderr
        );
    }
    Ok(())
}

/// Splits git output into lines, trims whitespace, and filters out
/// empty lines and node_modules paths. Matches workspace-tools processGitOutput
/// with excludeNodeModules: true.
fn process_git_output(stdout: &str) -> Vec<String> {
    stdout
        .lines()
        .map(str::trim)
        .filter(|l| !l.is_empty() && !l.contains("node_modules"))
        .map(str::to_string)
        .collect()
}

File: go/internal/git/commands.go
package git

import (
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"slices"
	"strings"

	"github.com/microsoft/beachball/internal/logging"
)

// GitResult holds the result of a git command.
type GitResult struct {
	Success  bool
	Stdout   string
	Stderr   string
	ExitCode int
}

// Git runs a git command in the given directory.
func Git(args []string, cwd string) (GitResult, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd

	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return GitResult{}, fmt.Errorf("failed to run git %s: %w", strings.Join(args, " "), err)
		}
	}

	result := GitResult{
		Success:  exitCode == 0,
		Stdout:   strings.TrimSpace(stdout.String()),
		Stderr:   strings.TrimSpace(stderr.String()),
		ExitCode: exitCode,
	}
	return result, nil
}

// gitStdout runs a git command and returns stdout, or error if it fails.
func gitStdout(args []string, cwd string) (string, error) {
	result, err := Git(args, cwd)
	if err != nil {
		return "", err
	}
	if !result.Success {
		return "", fmt.Errorf("git %s failed: %s", strings.Join(args, " "), result.Stderr)
	}
	return result.Stdout, nil
}

// FindGitRoot returns the root directory of the git repository.
func FindGitRoot(cwd string) (string, error) {
	return gitStdout([]string{"rev-parse", "--show-toplevel"}, cwd)
}

// managerFiles are workspace/monorepo manager config files, in precedence order.
// Matches the workspace-tools detection order.
var ManagerFiles = []string{
	"lerna.json",
	"rush.json",
	"yarn.lock",
	"pnpm-workspace.yaml",
	"package-lock.json",
}

// searchUp walks up the directory tree from cwd looking for any of the given files.
// Returns the full path of the first match, or "" if none found.
func searchUp(files []string, cwd string) string {
	absPath, err := filepath.Abs(cwd)
	if err != nil {
		return ""
	}
	root := filepath.VolumeName(absPath) + string(filepath.Separator)

	dir := absPath
	for dir != root {
		for _, f := range files {
			candidate := filepath.Join(dir, f)
			if _, err := os.Stat(candidate); err == nil {
				return candidate
			}
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	return ""
}

// FindProjectRoot searches up from cwd for a workspace manager root,
// falling back to the git root. Matches workspace-tools findProjectRoot.
func FindProjectRoot(cwd string) (string, error) {
	if found := searchUp(ManagerFiles, cwd); found != "" {
		return filepath.Dir(found), nil
	}
	return FindGitRoot(cwd)
}

// GetBranchName returns the current branch name.
func GetBranchName(cwd string) (string, error) {
	return gitStdout([]string{"rev-parse", "--abbrev-ref", "HEAD"}, cwd)
}

// GetUserEmail returns the user's git email.
func GetUserEmail(cwd string) string {
	email, err := gitStdout([]string{"config", "user.email"}, cwd)
	if err != nil {
		return ""
	}
	return email
}

// GetBranchChanges returns files changed between the current branch and the target branch.
func GetBranchChanges(branch, cwd string) ([]string, error) {
	result, err := Git([]string{
		"--no-pager", "diff", "--name-only", "--relative",
		fmt.Sprintf("%s...", branch),
	}, cwd)
	if err != nil {
		return nil, err
	}
	if !result.Success {
		return nil, nil
	}
	return processGitOutput(result.Stdout), nil
}

// GetStagedChanges returns staged file changes.
func GetStagedChanges(cwd string) ([]string, error) {
	result, err := Git([]string{
		"--no-pager", "diff", "--staged", "--name-only", "--relative",
	}, cwd)
	if err != nil {
		return nil, err
	}
	if !result.Success {
		return nil, nil
	}
	return processGitOutput(result.Stdout), nil
}

// GetChangesBetweenRefs returns changes between refs with optional filter and pattern.
func GetChangesBetweenRefs(fromRef string, diffFilter, pattern, cwd string) ([]string, error) {
	args := []string{"--no-pager", "diff", "--name-only", "--relative"}
	if diffFilter != "" {
		args = append(args, fmt.Sprintf("--diff-filter=%s", diffFilter))
	}
	args = append(args, fmt.Sprintf("%s...", fromRef))
	if pattern != "" {
		args = append(args, "--", pattern)
	}

	result, err := Git(args, cwd)
	if err != nil {
		return nil, err
	}
	if !result.Success {
		return nil, nil
	}
	return processGitOutput(result.Stdout), nil
}

// GetUntrackedChanges returns untracked files (not in index, respecting .gitignore).
func GetUntrackedChanges(cwd string) ([]string, error) {
	result, err := Git([]string{"ls-files", "--others", "--exclude-standard"}, cwd)
	if err != nil {
		return nil, err
	}
	if !result.Success {
		return nil, nil
	}
	return processGitOutput(result.Stdout), nil
}

// Stage adds files to the staging area.
func Stage(files []string, cwd string) error {
	args := append([]string{"add"}, files...)
	_, err := gitStdout(args, cwd)
	return err
}

// Commit creates a commit with the given message.
func Commit(message, cwd string) error {
	_, err := gitStdout([]string{"commit", "-m", message}, cwd)
	return err
}

// IsShallowClone checks if the repo is a shallow clone.
func IsShallowClone(cwd string) bool {
	result, err := gitStdout([]string{"rev-parse", "--is-shallow-repository"}, cwd)
	if err != nil {
		return false
	}
	return result == "true"
}

// Fetch fetches from the remote branch.
func Fetch(remoteBranch, cwd string) error {
	parts := strings.SplitN(remoteBranch, "/", 2)
	if len(parts) != 2 {
		return fmt.Errorf("invalid remote branch format: %s", remoteBranch)
	}
	_, err := gitStdout([]string{"fetch", "--", parts[0], parts[1]}, cwd)
	return err
}

// Deepen deepens a shallow clone.
func Deepen(depth int, cwd string) error {
	_, err := gitStdout([]string{"fetch", "--deepen", fmt.Sprintf("%d", depth)}, cwd)
	return err
}

// HasRef checks if a ref exists.
func HasRef(ref, cwd string) bool {
	result, err := Git([]string{"rev-parse", "--verify", ref}, cwd)
	if err != nil {
		return false
	}
	return result.Success
}

// GetDefaultRemote returns the default remote name, matching workspace-tools getDefaultRemote.
//
// The order of preference is:
//  1. If `repository` is defined in package.json at the git root, the remote with a matching URL
//  2. "upstream" if defined
//  3. "origin" if defined
//  4. The first defined remote
//  5. "origin" as final fallback
//
// Note: ADO/VSO URL formats are not currently handled by getRepositoryName.
// This is probably fine since usage of forks with ADO is uncommon.
func GetDefaultRemote(cwd string) string {
	gitRoot, err := FindGitRoot(cwd)
	if err != nil {
		return "origin"
	}

	// Read package.json at git root for repository field
	repositoryName := ""
	packageJsonPath := filepath.Join(gitRoot, "package.json")
	pkgData, err := os.ReadFile(packageJsonPath)
	if err != nil {
		logging.Info.Printf(`Valid "repository" key not found in "%s". Consider adding this info for more accurate git remote detection.`, packageJsonPath)
	} else {
		var pkg struct {
			Repository json.RawMessage `json:"repository"`
		}
		if err := json.Unmarshal(pkgData, &pkg); err == nil && pkg.Repository != nil {
			repositoryURL := extractRepositoryURL(pkg.Repository)
			if repositoryURL == "" {
				logging.Info.Printf(`Valid "repository" key not found in "%s". Consider adding this info for more accurate git remote detection.`, packageJsonPath)
			} else {
				repositoryName = getRepositoryName(repositoryURL)
			}
		} else {
			logging.Info.Printf(`Valid "repository" key not found in "%s". Consider adding this info for more accurate git remote detection.`, packageJsonPath)
		}
	}

	// Get remotes with URLs
	remotesResult, err := Git([]string{"remote", "-v"}, cwd)
	if err != nil || !remotesResult.Success {
		return "origin"
	}

	// Build mapping from repository name → remote name
	remotesByRepoName := map[string]string{}
	var allRemoteNames []string
	seen := map[string]bool{}
	for _, line := range splitLines(remotesResult.Stdout) {
		fields := strings.Fields(line)
		if len(fields) < 2 {
			continue
		}
		remoteName, remoteURL := fields[0], fields[1]
		repoName := getRepositoryName(remoteURL)
		if repoName != "" {
			remotesByRepoName[repoName] = remoteName
		}
		if !seen[remoteName] {
			allRemoteNames = append(allRemoteNames, remoteName)
			seen[remoteName] = true
		}
	}

	// 1. Match by repository name from package.json
	if repositoryName != "" {
		if matched, ok := remotesByRepoName[repositoryName]; ok {
			return matched
		}
	}

	// 2-4. Fall back to upstream > origin > first
	if slices.Contains(allRemoteNames, "upstream") {
		return "upstream"
	}
	if slices.Contains(allRemoteNames, "origin") {
		return "origin"
	}
	if len(allRemoteNames) > 0 {
		return allRemoteNames[0]
	}

	return "origin"
}

// extractRepositoryURL gets the URL from a package.json "repository" field,
// which can be a string or an object with a "url" property.
func extractRepositoryURL(raw json.RawMessage) string {
	// Try as string first
	var s string
	if json.Unmarshal(raw, &s) == nil && s != "" {
		return s
	}
	// Try as object with url field
	var obj struct {
		URL string `json:"url"`
	}
	if json.Unmarshal(raw, &obj) == nil {
		return obj.URL
	}
	return ""
}

// sshPattern matches SSH git URLs like git@github.com:owner/repo.git
var sshPattern = regexp.MustCompile(`^[^@]+@([^:]+):(.+?)(?:\.git)?$`)

// shorthandPattern matches shorthand URLs like github:owner/repo
var shorthandPattern = regexp.MustCompile(`^[a-z]+:([^/].+)$`)

// getRepositoryName extracts the "owner/repo" full name from a git URL.
// Handles HTTPS, SSH, git://, and shorthand (github:owner/repo) formats for
// common hosts (GitHub, GitLab, Bitbucket, etc.).
//
// Note: Azure DevOps and Visual Studio Online URL formats are not currently handled.
// Those would require more complex parsing similar to workspace-tools' git-url-parse usage.
func getRepositoryName(rawURL string) string {
	if rawURL == "" {
		return ""
	}

	// SSH format: git@github.com:owner/repo.git
	if m := sshPattern.FindStringSubmatch(rawURL); m != nil {
		return strings.TrimSuffix(m[2], ".git")
	}

	// Shorthand format: github:owner/repo
	if m := shorthandPattern.FindStringSubmatch(rawURL); m != nil {
		// e.g. "github:microsoft/workspace-tools" → "microsoft/workspace-tools"
		return strings.TrimSuffix(m[1], ".git")
	}

	// HTTPS or git:// format
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return ""
	}
	path := strings.TrimPrefix(parsed.Path, "/")
	path = strings.TrimSuffix(path, ".git")
	if path == "" {
		return ""
	}
	return path
}

// GetDefaultRemoteBranch returns the default remote branch reference (e.g. "origin/main").
// If branch is non-empty, it's combined with the default remote.
// If branch is empty, detects the remote's HEAD branch, falling back to
// git config init.defaultBranch or "master".
func GetDefaultRemoteBranch(cwd string, branch string) string {
	remote := GetDefaultRemote(cwd)

	if branch != "" {
		return remote + "/" + branch
	}

	// Try to detect HEAD branch from remote
	result, err := Git([]string{"remote", "show", remote}, cwd)
	if err == nil && result.Success {
		for line := range strings.SplitSeq(result.Stdout, "\n") {
			trimmed := strings.TrimSpace(line)
			if after, ok := strings.CutPrefix(trimmed, "HEAD branch:"); ok {
				headBranch := strings.TrimSpace(after)
				return remote + "/" + headBranch
			}
		}
	}

	// Fallback: git config init.defaultBranch, or "master"
	if defaultBranch, err := gitStdout([]string{"config", "init.defaultBranch"}, cwd); err == nil && defaultBranch != "" {
		return remote + "/" + defaultBranch
	}

	return remote + "/master"
}

// processGitOutput splits git output into lines, trims whitespace, and filters out
// empty lines and node_modules paths. Matches workspace-tools processGitOutput with
// excludeNodeModules: true.
func processGitOutput(s string) []string {
	if s == "" {
		return nil
	}
	var lines []string
	for line := range strings.SplitSeq(s, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" && !strings.Contains(trimmed, "node_modules") {
			lines = append(lines, trimmed)
		}
	}
	return lines
}

// splitLines splits output into non-empty lines (without node_modules filtering).
// Used for non-file-path output like remote names.
func splitLines(s string) []string {
	if s == "" {
		return nil
	}
	var lines []string
	for line := range strings.SplitSeq(s, "\n") {
		if line != "" {
			lines = append(lines, line)
		}
	}
	return lines
}
